teamcity: failed test: TestLogic

The following tests appear to have failed on master (test): TestLogic/fakedist: TestLogic/fakedist-metadata, TestLogic, TestLogic/fakedist: TestLogic/fakedist-metadata: TestLogic/fakedist-metadata/fk

You may want to check [for open issues](https://github.com/cockroachdb/cockroach/issues?q=is%3Aissue+is%3Aopen+TestLogic).

[#1156962](https://teamcity.cockroachdb.com/viewLog.html?buildId=1156962):


```
TestLogic/fakedist: TestLogic/fakedist-metadata
--- FAIL: test/TestLogic: TestLogic/fakedist: TestLogic/fakedist-metadata (0.360s)




TestLogic/fakedist: TestLogic/fakedist-metadata: TestLogic/fakedist-metadata/fk
--- FAIL: test/TestLogic: TestLogic/fakedist: TestLogic/fakedist-metadata: TestLogic/fakedist-metadata/fk (2.540s)

------- Stdout: -------
=== PAUSE TestLogic/fakedist-metadata/fk
            logic.go:2313: 
                 
                testdata/logic_test/fk:318: 
                expected "foreign key violation: \"delivery\" row item='885155001450' has no match in \"products\"", but found "pq: expected 13 metadata senders but found 11; missing ae10e1a7-4615-45ed-beb8-fb4632e27be3, 5436c8c4-3b88-4ed4-ac49-a2e3d2c5b392"
            logic.go:2345: 
                testdata/logic_test/fk:321: error while processing
            logic.go:2346: testdata/logic_test/fk:321: too many errors encountered, skipping the rest of the input



TestLogic
--- FAIL: test/TestLogic (422.610s)

------- Stdout: -------
    test_log_scope.go:81: test logs captured to: /tmp/logTestLogic876114019
    test_log_scope.go:62: use -show-logs to present logs inline




```

Please assign, take a look and update the issue accordingly.

--- a/pkg/sql/distsqlrun/input_sync.go
+++ b/pkg/sql/distsqlrun/input_sync.go
@@ -71,8 +71,10 @@ type orderedSynchronizer struct {
 	// state dictates the operation mode.
 	state orderedSynchronizerState
 
-	// heap of source indexes, ordered by the current row. Sources with no more
-	// rows are not in the heap.
+	// heap of source indexes. In state notInitialized, heap holds all source
+	// indexes. Once initialized (initHeap is called), heap will be ordered by the
+	// current row from each source and will only contain source indexes of
+	// sources that are not done.
 	heap []srcIdx
 	// needsAdvance is set when the row at the root of the heap has already been
 	// consumed and thus producing a new row requires the root to be advanced.
@@ -139,21 +141,24 @@ func (s *orderedSynchronizer) Pop() interface{} {
 func (s *orderedSynchronizer) initHeap() error {
 	// consumeErr is the last error encountered while consuming metadata.
 	var consumeErr error
-	for i := range s.sources {
-		src := &s.sources[i]
+
+	// s.heap contains all the sources. Go through the sources and get the first
+	// row from each source, removing them if they're already finished.
+	toDelete := 0
+	for i, srcIdx := range s.heap {
+		src := &s.sources[srcIdx]
 		err := s.consumeMetadata(src, stopOnRowOrError)
 		if err != nil {
 			consumeErr = err
 		}
-		// We add the source to the heap either if we have received a row from
-		// it or there was an error reading from this source. We still add to
-		// the heap in case of error so that these sources can be drained in
-		// `drainSources`.
-		if src.row != nil || err != nil {
-			// Add to the heap array (it won't be a heap until we call heap.Init).
-			s.heap = append(s.heap, srcIdx(i))
+		if src.row == nil && err == nil {
+			// Source is done. Swap current element with the first of the valid
+			// sources then truncate array after the loop so that i is still valid.
+			s.heap[toDelete], s.heap[i] = s.heap[i], s.heap[toDelete]
+			toDelete++
 		}
 	}
+	s.heap = s.heap[toDelete:]
 	if consumeErr != nil {
 		return consumeErr
 	}
@@ -366,6 +371,7 @@ func makeOrderedSync(
 	}
 	for i := range s.sources {
 		s.sources[i].src = sources[i]
+		s.heap = append(s.heap, srcIdx(i))
 	}
 	return s, nil
 }

--- a/pkg/sql/distsqlrun/input_sync_test.go
+++ b/pkg/sql/distsqlrun/input_sync_test.go
@@ -140,6 +140,48 @@ func TestOrderedSync(t *testing.T) {
 	}
 }
 
+func TestOrderedSyncDrainBeforeNext(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	expectedMeta := &ProducerMetadata{Err: errors.New("expected metadata")}
+
+	var sources []RowSource
+	for i := 0; i < 4; i++ {
+		rowBuf := NewRowBuffer(sqlbase.OneIntCol, nil /* rows */, RowBufferArgs{})
+		sources = append(sources, rowBuf)
+		rowBuf.Push(nil, expectedMeta)
+	}
+
+	ctx := context.Background()
+	evalCtx := tree.NewTestingEvalContext(cluster.MakeTestingClusterSettings())
+	defer evalCtx.Stop(ctx)
+	o, err := makeOrderedSync(sqlbase.ColumnOrdering{}, evalCtx, sources)
+	if err != nil {
+		t.Fatal(err)
+	}
+	o.Start(ctx)
+
+	// Call ConsumerDone before Next has been called.
+	o.ConsumerDone()
+
+	metasFound := 0
+	for {
+		_, meta := o.Next()
+		if meta == nil {
+			break
+		}
+
+		if meta != expectedMeta {
+			t.Fatalf("unexpected meta %v, expected %v", meta, expectedMeta)
+		}
+
+		metasFound++
+	}
+	if metasFound != len(sources) {
+		t.Fatalf("unexpected number of metadata items %d, expected %d", metasFound, len(sources))
+	}
+}
+
 func TestUnorderedSync(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 

